python/lsst/daf/butler/core/registry.py

```diff
@@ -76,6 +78,185 @@ class RegistryConfig(ConfigSubset):
+        if parsed.scheme and not parsed.netloc:
+            # driver is infered from target registry class when not explicit
+            if '+' in parsed.scheme:
+                conStrKwargs['dialect'], conStrKwargs['driver'] = parsed.scheme.split('+')
+            else:
+                conStrKwargs['driver'] = registryCls.driver
+
+            # when dialects match use it, when they don't 'db' is an URI, URL
+            # (handled later) or inherited from default (handled now).
+            # Everything else is resolved according to urllib.
+            if registryCls.dialect == parsed.scheme:
+                conStrKwargs['dialect'] = parsed.scheme
+            elif registryCls.dialect != parsed.scheme:
+                conStrKwargs['dialect'] = registryCls.dialect
+                conStrKwargs['driver'] = registryCls.driver
+            else:
+                conStrKwargs['dialect'] = parsed.scheme
+                conStrKwargs['dbname'] = parsed.path
+
+            conStrKwargs['dbname'] = parsed.path
+            conStrKwargs['port'] = parsed.port
+
+        # scheme-less URLs (f.e. host.com(:port?)(/dbname?)) or its sqlite so
+        # it's safe to ignore it (all cases handled by first or last section)
+        if (not parsed.scheme and not parsed.netloc) or 'sqlite' not in parsed.scheme:
+            tmpparse = urlparse.urlparse('http://'+regConf['db'])
+            conStrKwargs['dialect'] = registryCls.dialect
+            conStrKwargs['driver'] = registryCls.driver
+            conStrKwargs['host'] = tmpparse.hostname
+            conStrKwargs['port'] = tmpparse.port
+            conStrKwargs['dbname'] = tmpparse.path
+
+        # fully qualified URI
+        if parsed.scheme and parsed.netloc:
+            if '+' in parsed.scheme:
+                conStrKwargs['dialect'], conStrKwargs['driver'] = parsed.scheme.split('+')
+            else:
+                conStrKwargs['dialect'] = parsed.scheme
+                conStrKwargs['driver'] = registryCls.driver
+            conStrKwargs['host'] = parsed.hostname
+            conStrKwargs['port'] = parsed.port
+            conStrKwargs['dbname'] = parsed.path
+
+        # are explicit plain text passwords in config files allowed?
+        # I like the ability to at least target a specific user explicitly
+        conStrKwargs['username'] = parsed.username
+        conStrKwargs['password'] = parsed.password
+
+        #
+        #    Merge the information from the two dictionaries.
+        #
+        mergedKwargs = dict()
+        for key, val in explicitKwargs.items():
+            mergedKwargs[key] = val
+            if val is None:
+                mergedKwargs[key] = conStrKwargs[key]
+
+        return mergedKwargs
+
+    @classmethod
+    def fromConfig(cls, config, withPass=True):
+        conKwargs = cls._dictFromRegistryConf(config)
+        validationKeys = list(cls.requiredKeys)
+        constr = "{dialect}+{driver}://{username}:{password}@{host}:{port}{dbname}"
```

Comment:
Partially, `DBAuth` matches on `host` and `port` so in some cases `port` becomes mandatory. I don't believe in the previous case it would have been rendered as "None" but as "". This was used in cases where `dbname` was the path to a file as then it would, usually, get shifted leftwards so file-targeting DB connection strings would work too. 

In any case, not important in the reworked version.